### PR TITLE
Jepsen test plan

### PR DESCRIPTION
**Pull request by aphyr**
A brief sketch of a plan for testing Materialize with Jepsen.

### Motivation

This plan exists to remind me what I'll be doing when I come back to test Materialize later, and to hint to Materialize's engineers about what kind of high-level API calls we'll be making, and what properties we'll verify.

### Testing

This is a planning document; no code is involved.

### Release notes

There is no user-facing impact

### CODE REVIEW

**Review by benesch — APPROVED** [review]
This all makes sense to me! I've requested review from @philip-stoev and @frankmcsherry as well.

**Review by philip-stoev — APPROVED** [review]
Thank you. Are we going to also formalize at this stage the failure modes that will be tested?

**Comment by aphyr** [comment]
Sure! I sketched out some ideas around which kinds of faults we might test, and the order we could go about testing them in.

**Comment by philip-stoev** [comment]
I have a couple of questions:

1. Is a compare-and-set implementation required for Elle to function? If so, how would that be implemented in Kafka?

2. If ingestion into Mz stops altogether, would that be caught by an Elle-based test? That is, if all reads on the Mz side always return e.g. `T0: read(x, {0})` as per the notation in the paper, would that be considered a consistent read?

3. Currently, durable user tables in Mz are limited to one update for every timestamp interval, which by default is 1s but can be brought down to 100ms and maybe 10ms if we are pushing it. Will 10 to 100 transactions per second provide a sufficiently current workload for Elle to detect anomalies, or we need to lift the limitation before any meaningful testing with crashes can happen?

4. Kafka and Postgres transactions arrive bunched in Mz, that is, if the original had 2 independent transactions bringing the system in discrete steps from state S0 through state S1 to state S2, what Mz will see is a direct jump from S0 to S2 if the two transactions were ingested within the timestamp interval (default is 1 second, can be lowered) . Will that loss of granularity/information be a problem? If Jepsen is hooked to perform reads from Mz, it will never see the effects of S1 alone on the Mz side. It will see the affects of S1 and S2 concatenated and possibly compacted together into one if a sink or tail is used to send the information out of Mz.

5. Would it be possible to also test `TAIL` for correctness? Subscribing to receive TAIL updates will give you a stream that looks like this:
```
...
1647342211227	-1	123	0,1
1647342211227	1	123	0,1,2
...
```
In other words, at timestamp 1647342211227 we retracted a tuple `(id=123, value=[0,1])` and produced a tuple `(id=123, value=[0,1,2])`. All the events emitted at a given timestamp form a single transaction of sorts. The `+1` entries could be considered "writes", while the `-1` entries either ignored or considered the reads(?)

6. In addition to `SELECT` and `TAIL`, information can be exfiltrated out of Mz by means of a sink, that is, by sending the ouput of the dataflow to a Kafka topic. Would it be possible to somehow validate sinks as well? In addition to emitting correct updates, for sinks we are also concerned about failing to emit altogether and emitting twice (especially on restart).
